validator_client/src/duties_service.rs

```diff
@@ -671,6 +691,35 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
             .data
             .into_iter()
             .filter(|duty| {
+                if duties_service.per_validator_metrics() {
+                    let validator_index = duty.validator_index;
+                    let duty_slot = duty.slot;
+                    if let Some(existing_slot_gauge) =
+                        get_int_gauge(&ATTESTATION_DUTY, &[&validator_index.to_string()])
+                    {
+                        let existing_slot = Slot::new(existing_slot_gauge.get() as u64);
+                        let existing_epoch = existing_slot.epoch(E::slots_per_epoch());
+
+                        // First condition ensures that we switch to the next epoch duty slot
+                        // once the current epoch duty slot passes.
+                        // Second condition is to ensure that next epoch duties don't override
+                        // current epoch duties.
```

Comment:
```suggestion
                        // Second condition is to ensure that current epoch duties override
                        // next epoch duties if provided out of order, or duties get updated by a re-org.
```

Is my understanding here correct?

If we see duties for a validator at slots `[35, 5]` before slot 5 then we'll set `existing_slot == 35` when the first duty is processed. Then when the second duty is found we'll override and set the metric to 5, because the 2nd duty slot is from an earlier or equal epoch, and it is yet to happen (`duty_slot > current_slot`).

Most of the time this won't happen though because we'll loop through current epoch duties before next epoch duties. So we'll process `[5, 35]`, in which case `5` will persist until it is in the past.

For re-orgs, the 2nd disjunct also handles them by allowing an override at the same epoch for a different slot that hasn't happened yet.

I think I'm happy with that! It made me think though! :sweat_smile:

Reply:
I hadn't thought about the duties coming out of order case tbh but looks like it works out.